stan/math/gpu/copy.hpp

```diff
@@ -17,7 +19,56 @@ namespace stan {
 namespace math {
 
 /**
- * Copies the source Eigen matrix to
+ * Copies the source primitive type Eigen matrix to
+ * the destination matrix that is stored
+ * on the GPU.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param dst destination matrix on the GPU
+ * @param src source Eigen matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ */
+template <typename T, int R, int C,
+          typename = typename std::enable_if<is_constant<T>::value, T>::type>
+void copy(matrix_gpu& dst, const Eigen::Matrix<T, R, C>& src) {
```

Comment:
What does this implementation do for types that aren't double that we might reasonably expect here? (e.g. `int`, `fvar<fvar<var>>`, `float`). Seems like it might do the right thing for some but not others.

Reply:
int and float would definitely pass through here. I'm not sure if that's good. Maybe this should only be for doubles?

Since `is_constant` checks if the type can be converted to a double `fvar<...>` would fail here right?

Reply:
I guess I'm wondering if you wanted it to work with ints, floats, and doubles, and if so we should probably make sure we test each. I think that's how `is_constant` should work. That seems reasonable.

Reply:
I think it's because most of the kernels assume we are receiving a double. I'm going to change this to only take in doubles